changing curwin in callback results in scrolling for 'nosplitscroll'

luukvbaal opened the pull request:
changing curwin in callback results in scrolling for 'nosplitscroll'

Problem: w_valid flags are not updated accordingly for 'nosplitscroll', missing test for terminal mode callback.
Solution: Let validate_cursor handle w_valid flags except for VALID_TOPLINE, add callback tests.

This fixes scrolling when the current window is changed inside a callback as I alluded to in: https://github.com/vim/vim/pull/11166
> Moreover, if we move the wincmd k to the callback function, the current fix is not enough to catch that and scrolling will still happen. Trying to fix it with a watchpoint on firstwin->w_topline and backtracing led me to trying to add workarounds in redraw_after_callback() or ex_docmd.c but it got quite complicated. Not sure if it's worth to catch this issue.

If we now catch all possible cases that could result in scrolling due to a user script callback I'm fine with this patch. I however can't be 100% certain that that is the case. I did try various other split/window commands inside the callback script and the added test keeps passing. I'm sure some user will report this issue at some point so up to you if you think this is a valid patch @brammool. Also adds missing test for the terminal mode callback patch added previously.

zeertzjq commented on src/testdir/test_window_cmd.vim:
```diff
@@ -1817,4 +1817,35 @@ function Test_nosplitscroll_misc()
   set splitscroll&
 endfunc
 
+source screendump.vim
```
This is usually put at the top at the file.

luukvbaal commented on src/window.c:
```diff
@@ -6383,6 +6383,7 @@ win_fix_scroll(int resize)
 		wp->w_cursor.lnum = lnum;
 	    }
 	    invalidate_botline_win(wp);
+	    wp->w_valid &= ~VALID_CROW;
```
Simply invalidating VALID_CROW fixes the issue. But since invalidating it is only necessary for this callback case, I wonder if there is a better place to invalidate it without needlessly making comp_botline() more expensive.

luukvbaal replied:
At the very least we could guard it behind get_callback_depth(), but this still indicates to me that we need to invalidate it somewhere else.

luukvbaal commented on src/window.c:
```diff
@@ -6382,6 +6382,8 @@ win_fix_scroll(int resize)
 	    }
+	    if (wp == curwin)
+		wp->w_valid &= ~VALID_CROW;
 	    invalidate_botline_win(wp);
```
I couldn't figure out a way to re-use validate_cursor()/check_cursor_moved(). This is the call we are skipping in
win_new_height() that is responsible for invalidating w_crow but using that unconditionally in win_fix_scroll/cursor results in scrolling. Invalidating it only seems to be necessary for curwin, and not invalidating it seems to cause an issue only in this callback case. Not sure if it makes sense to also guard behind get_callback_depth().

luukvbaal committed:
changing curwin in callback results in scrolling for 'nosplitscroll'

Problem:	Changing curwin in callback results in scrolling
for 'nosplitscroll', missing test for terminal mode callback.
Solution:	Invalidate w_crow for curwin, add callback tests.
